Bump Spectre.Console, refactor Format-SpectreTable, fix bugs, and add new features and tests

trackd opened the pull request:
### Summary

* Bump Spectre.Console from version 0.47.0 to 0.48.0
* Added new public function, `Write-SpectreCalendar`
* `Format-SpectreTable`
  * fix bugs in #20
  * added functionality for enumerating scalar arrays #20 
  * refactored into smaller, more manageable functions for easier testing and verification
    * `Add-TableColumn`
    * `New-TableCell`
    * `New-TableRow`
    * `Test-IsScalar`
    * `ConvertFrom-ConsoleColor`
* Fix Console Color conversion (4-bit)
* `ConvertTo-SpectreDecoration`, better cleaning of strings.
* Cleanup of `Get-DefaultDisplayMembers`
* Fix names in decoration class
* improved testhelpers
  * New mock function to capture rendered Spectre objects, `Get-SpectreRenderable`
  * `Get-AnsiEscapeSequence`
  * `Get-PSStyleRandom`
  * `Get-SpectreProfile`
  * `Get-SpectreColorSample`
* added more tests

ShaunLawrie commented:
> cleanup Get-DefaultDisplayMembers (should we change that name? could be something shorter like Get-TableFormat)

I'm not really fussed on private function names being a bit verbose, it seems to describe what it does quite well the way it is. Do whatever you want with it though, I appreciate the work you've done on this 🙏

ShaunLawrie commented:
Most of the logic is in the column and row wrangling, it might be better to handle them independently i.e. set up the columns fully first then set up the row data. It could be broken out into a couple of functions like below to make the `end {}` block easier to read + the bits could be tested independently.

👇 this is pseudo-code as heck, I might have missed some things that make this not work the way I think it will...

**End block:**
```pwsh
end {
  if("nothing is in the collector") {
    # bail out
  }
  
  $standardMembers = Get-DefaultDisplayMembers # add the check here for .PSTypeNames[0] -notmatch 'PSCustomObject'
  
  Add-TableColumns -Property $Property -StandardMembers $StandardMembers -Table $table -Collector $Collector
  
  if($Property) {
    $collector = $collector | Select-Object -Property $Property
  } elseif($standardMembers) {
    $collector = $collector | Select-Object $standardMembers.Format
  }
  
  foreach ($item in $collector) {
    $row = New-TableRow -Item $item -AllowMarkup $AllowMarkup
    [TableExtensions]::AddRow($table, $row) # this works without casting the array of rows so no AllowMarkup check required
  }
  
  Write-AnsiConsole $table
}
```

**Functions supporting the above:**
```pwsh
# This is the bulk of the existing code in the end block
function Add-TableColumns {
  param (
    $property,
    $standardMembers
    $table,
    $collector
  )
  if("user requested a specific list of $property") {
    # select them from the items in the $collector
    # add them to the $table
  } elseif("$standardMembers are available / not null") {
    # add the columns to the $table with available formatting
  } elseif(Test-IsScalar $collector[0].PSObject) {
    # simple/scalar types show up wonky, we can detect them and just use a dummy header for the table
    $table.AddColumn("Value")
  } else {
    # no formatting found and no properties selected, enumerating psobject.properties.name
  }
}
```
```pwsh
# Check if the object type is scalar/simple
function Test-IsScalar {
  param (
    $Value
  )
  $matchedScalarTypes = $Value.PSObject.TypeNames | Where-Object {
    @("System.String", "System.ValueType") -contains $_ 
  }
  return $null -ne $matchedScalarTypes
}
```
```pwsh
# I should have done this when I smushed the AllowMarkup stuff into your branch, handling the allowmarkup inline makes the "end {}" block huge
function New-TableRow {
  param (
    [object] $Item,
    [boolean] $AllowMarkup
  )
  $columns = @()
  foreach($cell in $item.PSObject.Properties) {
    if([string]::IsNullOrEmpty($cell.Value)) {
      $Value = " "
    }
    if($AllowMarkup) {
      $columns += [Markup]::new([string]$cell.Value)
    } else {
      $columns += [Text]::new([string]$cell.Value)
    }
  }
  return $columns
}
```

ShaunLawrie commented:
The default behaviour for dumping an array of hashtables to a spectre table could also be cast to a pscustom object in the process block. I'm not sure of many situations where I would want to dump the properties of the hashtable vs it trying to output the actual hashtable entries
```pwsh
foreach ($entry in $data) {
    if($entry -is [hashtable]) {
        $collector.add([pscustomobject]$entry)
    } else {
        $collector.add($entry)
    }
}
```
<img width="750" alt="image" src="https://github.com/ShaunLawrie/PwshSpectreConsole/assets/13159458/7922c386-f413-42ec-ae59-8121e6017c3f">

trackd commented:
Really like the proposed changes!
I'll try and add this to the pr, during the weekend.

fmotion1 commented:
Proposed changes look great IMO as well! Looking forward to trying it all out when implemented.

trackd commented:
this test here works for me, im not sure if it covers all scenarios..

(and ofcourse the pester tests)

did i miss any object type we should render?

I need to do some cleanup before merge, theres alot of Write-Debug stuff in the code at the moment.
but prefer to keep it in while working on it.

```ps1
[System.Collections.Generic.List[object]]$GenericList = @()
1..5 | ForEach-Object { $GenericList.Add([PSCustomObject]@{ Name = (Get-Random) }) }
[System.Collections.Generic.List[System.IO.DirectoryInfo]]$Files = @()
$path = $env:PSModulePath -split ([System.IO.Path]::PathSeparator) | Select-Object -First 1
1..5 | ForEach-Object { $Files.Add([System.IO.DirectoryInfo]"$(Join-Path $path ([string](Get-Random) + '.tmp'))") }
$sample = @{
    StringArray = @("one", "two", "three")
    GenericList = $GenericList
    hashtable = @{
        PwshSpectreConsole = "nice"
        pwsh = $PSVersionTable.PSVersion
        third = 'something'
    }
    Filesystem = $Files
    PSCustomObject = [PSCustomObject]@{
        Name = 'PwshSpectre'
        Version = '0.0.1'
        Description = 'A PowerShell module to display data in a table'
        dummy = 'dummy'
        dummy2 = 'dummy2'
        dummy3 = 'dummy3'
    }
}

foreach ($test in $sample.GetEnumerator()) {
    Write-SpectreHost -Message "Testing $($test.Name)"
    if ($test.Name -eq 'hashtable') {
        Format-SpectreTable -Data $test.Value -Title 'My hashtable'
        continue
    }
    if ($Test.Name -eq 'PSCustomObject') {
        Format-SpectreTable -Data $test.Value -Title 'My PSCustomObject' -Property Name, Version, Description
        continue
    }
    Format-SpectreTable -Data $test.Value
}

```

ShaunLawrie commented:
There are some psstyle options I think will need to be mapped/stripped
based on it failing in one of the tests I just pushed.

I think I might have a couple of your scenarios in those tests, if I missed
some just copy paste the basic test and pipe in different input types.
For validation the best you can do is check the row counts are correct on
the returned table object because not much else is exposed publicly in
spectre console.

If you’re running the tests locally you need to run build.ps1 first to
download the spectre libs.

You can leave write debug in there. I find it convenient, I used to cmdlet
bind everything and pass $verbose preference the whole way through 🤦‍♂️
write-debug is much better.

On Mon, 8 Jan 2024 at 12:48 PM, trackd ***@***.***> wrote:

> this test here works for me, im not sure if it covers all scenarios..
>
> (and ofcourse the pester tests)
>
> did i miss any object type we should render?
>
> I need to do some cleanup before merge, theres alot of Write-Debug stuff
> in the code at the moment.
> but prefer to keep it in while working on it.
>
> [System.Collections.Generic.List[object]]$GenericList = @()1..5 | ForEach-Object { $GenericList.Add([PSCustomObject]@{ Name = (Get-Random) }) }
> [System.Collections.Generic.List[System.IO.DirectoryInfo]]$Files = @()$path = $env:PSModulePath -split ([System.IO.Path]::PathSeparator) | Select-Object -First 11..5 | ForEach-Object { $Files.Add([System.IO.DirectoryInfo]"$(Join-Path $path ([string](Get-Random) + '.tmp'))") }$sample = @{
>     StringArray = @("one", "two", "three")
>     GenericList = $GenericList
>     hashtable = @{
>         PwshSpectreConsole = "nice"
>         pwsh = $PSVersionTable.PSVersion
>         third = 'something'
>     }
>     Filesystem = $Files
>     PSCustomObject = [PSCustomObject]@{
>         Name = 'PwshSpectre'
>         Version = '0.0.1'
>         Description = 'A PowerShell module to display data in a table'
>         dummy = 'dummy'
>         dummy2 = 'dummy2'
>         dummy3 = 'dummy3'
>     }
> }
> foreach ($test in $sample.GetEnumerator()) {
>     Write-SpectreHost -Message "Testing $($test.Name)"
>     if ($test.Name -eq 'hashtable') {
>         Format-SpectreTable -Data $test.Value -Title 'My hashtable'
>         continue
>     }
>     if ($Test.Name -eq 'PSCustomObject') {
>         Format-SpectreTable -Data $test.Value -Title 'My PSCustomObject' -Property Name, Version, Description
>         continue
>     }
>     Format-SpectreTable -Data $test.Value
> }
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ShaunLawrie/PwshSpectreConsole/pull/21#issuecomment-1880222822>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADEMYIQPWLWR4L47BMH36KLYNMX27AVCNFSM6AAAAABBKXRFVKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOBQGIZDEOBSGI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

ShaunLawrie commented:
I’ve had to take off to an appointment sorry to push broken tests and bail
😂


On Mon, 8 Jan 2024 at 1:10 PM, Shaun Lawrie ***@***.***>
wrote:

> There are some psstyle options I think will need to be mapped/stripped
> based on it failing in one of the tests I just pushed.
>
> I think I might have a couple of your scenarios in those tests, if I
> missed some just copy paste the basic test and pipe in different input
> types.
> For validation the best you can do is check the row counts are correct on
> the returned table object because not much else is exposed publicly in
> spectre console.
>
> If you’re running the tests locally you need to run build.ps1 first to
> download the spectre libs.
>
> You can leave write debug in there. I find it convenient, I used to cmdlet
> bind everything and pass $verbose preference the whole way through 🤦‍♂️
> write-debug is much better.
>
> On Mon, 8 Jan 2024 at 12:48 PM, trackd ***@***.***> wrote:
>
>> this test here works for me, im not sure if it covers all scenarios..
>>
>> (and ofcourse the pester tests)
>>
>> did i miss any object type we should render?
>>
>> I need to do some cleanup before merge, theres alot of Write-Debug stuff
>> in the code at the moment.
>> but prefer to keep it in while working on it.
>>
>> [System.Collections.Generic.List[object]]$GenericList = @()1..5 | ForEach-Object { $GenericList.Add([PSCustomObject]@{ Name = (Get-Random) }) }
>> [System.Collections.Generic.List[System.IO.DirectoryInfo]]$Files = @()$path = $env:PSModulePath -split ([System.IO.Path]::PathSeparator) | Select-Object -First 11..5 | ForEach-Object { $Files.Add([System.IO.DirectoryInfo]"$(Join-Path $path ([string](Get-Random) + '.tmp'))") }$sample = @{
>>     StringArray = @("one", "two", "three")
>>     GenericList = $GenericList
>>     hashtable = @{
>>         PwshSpectreConsole = "nice"
>>         pwsh = $PSVersionTable.PSVersion
>>         third = 'something'
>>     }
>>     Filesystem = $Files
>>     PSCustomObject = [PSCustomObject]@{
>>         Name = 'PwshSpectre'
>>         Version = '0.0.1'
>>         Description = 'A PowerShell module to display data in a table'
>>         dummy = 'dummy'
>>         dummy2 = 'dummy2'
>>         dummy3 = 'dummy3'
>>     }
>> }
>> foreach ($test in $sample.GetEnumerator()) {
>>     Write-SpectreHost -Message "Testing $($test.Name)"
>>     if ($test.Name -eq 'hashtable') {
>>         Format-SpectreTable -Data $test.Value -Title 'My hashtable'
>>         continue
>>     }
>>     if ($Test.Name -eq 'PSCustomObject') {
>>         Format-SpectreTable -Data $test.Value -Title 'My PSCustomObject' -Property Name, Version, Description
>>         continue
>>     }
>>     Format-SpectreTable -Data $test.Value
>> }
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/ShaunLawrie/PwshSpectreConsole/pull/21#issuecomment-1880222822>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ADEMYIQPWLWR4L47BMH36KLYNMX27AVCNFSM6AAAAABBKXRFVKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOBQGIZDEOBSGI>
>> .
>> You are receiving this because you commented.Message ID:
>> ***@***.***>
>>
>

trackd commented:
> There are some psstyle options I think will need to be mapped/stripped based on it failing in one of the tests I just pushed.

I think i just copied values from [wikipedia/ANSI_escape_code](https://en.wikipedia.org/wiki/ANSI_escape_code#SGR_(Select_Graphic_Rendition)_parameters)

I'm sure we could tweak it a bit. those codes can be a bit spotty on the implementation side not all terminals use the same code for the same thing or implement all the codes.

hmm, looking at the test a bit i think we need to figure out a different way to test decorations.

`PSStyle` names dont match Spectre Names and we need to keep it as Spectre `Style` for it to work as a decoration.
[styles](https://spectreconsole.net/appendix/styles)
I did notice that i had wrong name for a few, gonna look them over and update the C# dictionary.

I think it would be better to just compare the ansi output if we could capture that and compare
maybe something like this would work?
```ps1
function Get-AnsiEscapeSequence {
    param(
        [Parameter(Mandatory, ValueFromPipeline)]
        $String
    )
    process {
        $decoded = $String.EnumerateRunes() | ForEach-Object {
            if ($_.Value -le 0x1f) {
                [Text.Rune]::new($_.Value + 0x2400)
            } else {
                $_
            }
        } | Join-String
        [PSCustomObject]@{
            Decoded  = $decoded 
            Original = $String
        }
    }
}
"$($PSStyle.blink) testblink $($PSStyle.Reset)" | Get-AnsiEscapeSequence
```
>You can leave write debug in there. I find it convenient, I used to cmdlet bind everything and pass $verbose preference the whole way through 🤦‍♂️ write-debug is much better.

it is _quite_ alot at the moment though, but ill leave it in.

trackd commented:
i looked over the TestHelpers module and added a new function, `Get-SpectreRenderable`

```ps1
        Describe "Get-SpectreRenderable" {
            It "returns the correct output for a given renderable" {
                $renderable = [Spectre.Console.Text]::new("Hello, world!")
                $output = Get-SpectreRenderable $renderable
                $output | Should -Be "Hello, world!"
            }
        }
        Describe "Get-SpectreRenderable" {
            It "returns the correct output for a given renderable" {
                $renderable = [Spectre.Console.Markup]::new("[Yellow]Hello[/], [Green]world![/]")
                $output = Get-SpectreRenderable $renderable
                $output | Should -Be "Hello, world!"
            }
        }

```

gives this,
![image](https://github.com/ShaunLawrie/PwshSpectreConsole/assets/17672644/12abf356-1557-488b-b562-dd57e8ed376b)

trackd commented:
also this package exists [Spectre.Console.Testing](https://www.nuget.org/packages/Spectre.Console.Testing/)
could look at implementing that, not to ship with the module ofcourse.

was looking at this as well
[Spectre Table tests](https://github.com/spectreconsole/spectre.console/tree/main/test/Spectre.Console.Tests/Expectations/Rendering/Borders/Table)

that's how spectre tests, should be possible to use `Get-SpectreRenderable` and compare with some saved .txt files but i havnt had the chance to test it fully

trackd commented on PwshSpectreConsole.Tests/formatting/new_Format-SpectreTable.tests.ps1:
```diff
@@ -9,6 +9,10 @@ Describe "Format-SpectreTable" {
             $testBorder = 'Markdown'
             $testColor = Get-RandomColor
             Mock Write-AnsiConsole {
+                param(
```
maybe try with param block,
```ps1
                param(
                    $RenderableObject
                )
```

ShaunLawrie replied:
Haha wow that pester test was annoying. It turns out the github action runner uses PowerShell 7.2 which doesn't have a parameter set that allows `-Skip` + `-SkipLast` in the same command. The github runners are just powershell on linux but I didn't think they'd be using the n-1 LTS version

ShaunLawrie replied:
They're using 7.2 https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/select-object?view=powershell-7.2

It looks like it has another test bug where the console width in the runner is set pretty small so the table column header being truncated with an ellipsis isn't being validated correctly.

ShaunLawrie commented:
> I havn't looked that closely at the build script, do i still need to run it or is it just the libs?
It just downloads the libs atm

trackd commented:
ok i figured out whats wrong with a few of the tests..
[testlog](https://github.com/trackd/PwshSpectreConsole/actions/runs/7467917117/job/20322370186)

added a small function `Get-SpectreProfile`

Seems the github runners are only running EightBit.

i tried setting $env:COLORTERM = 'truecolor' on a local linux machine but it doesn't seem to affect this,
[relevant code](https://github.com/spectreconsole/spectre.console/blob/main/src/Spectre.Console/Internal/Colors/ColorSystemDetector.cs#L40-L51)
a bit curious...

this means that spectre console will automatically downgrade any 24bit colors to a color that is supported, which is pretty cool but makes this test a bit hard.

but the tests work locally, and did find a big bug in my converter code that im gonna try and fix.

```ps1
Enrichers             : GitHub
ColorSystem           : EightBit
Unicode               : True
Ansi                  : True
Links                 : False
Legacy                : False
Interactive           : False
Terminal              : False
Writer                : System.IO.TextWriter+SyncTextWriter
Width                 : 80
Height                : 80
Encoding              : Unicode (UTF-8)
PSStyle               : Host
ConsoleOutputEncoding : System.Text.ConsoleEncoding
ConsoleInputEncoding  : System.Text.ConsoleEncoding
```

trackd commented:
we should probably disable a few of the tests for CI as it requires truecolor?

but all tests work locally for me atleast
```ps1
Tests completed in 13.12s
Tests Passed: 72, Failed: 0, Skipped: 0 NotRun: 0
Processing code coverage result.
Covered 43,46% / 75%. 1 560 analyzed Commands in 57 Files.
```

and all the examples from #20 work.

trackd commented:
now the function maps console colors correctly, but Spectre dont use the same escape sequence so its gonnna be a bit harder to verify in tests.

[Test-Colortable](https://gist.github.com/trackd/82d1355ec90e4722a20a2b72f21be453)
that's just for console colors.

it'll adapt to the colors you have chosen in your color profile in your terminal.
example
my normal profile using Nord color theme.
![image](https://github.com/ShaunLawrie/PwshSpectreConsole/assets/17672644/c1e02353-2464-4ac1-a486-fb69e1abfabc)
noprofile and using One Half Dark color theme
![image](https://github.com/ShaunLawrie/PwshSpectreConsole/assets/17672644/0706cf35-d587-4b33-b528-cbe333e2d0a0)

trackd commented:
@ShaunLawrie this is ready for prerelease.

I ran through `Start-SpectreDemo` and the Pester tests, all looks good.

CI tests failed but that's because they can only run 8bit so 24bit rgb is downgraded.


<details>
  <summary>Testlog</summary>
Starting pwsh -noprofile and Running Pester tests...

Starting discovery in 21 files.
Discovery found 72 tests in 2.84s.
Starting code coverage.
Running tests.
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\formatting\ConvertTo-SpectreDecoration.tests.ps1 1.94s (1.33s|296ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\formatting\Format-SpectreBarChart.tests.ps1 653ms (443ms|86ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\formatting\Format-SpectreBreakdownChart.tests.ps1 369ms (167ms|63ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\formatting\Format-SpectreJson.tests.ps1 302ms (127ms|46ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\formatting\Format-SpectrePanel.tests.ps1 215ms (57ms|40ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\formatting\Format-SpectreTable.tests.ps1 504ms (307ms|69ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\formatting\Format-SpectreTree.tests.ps1 234ms (76ms|41ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\formatting\new_Format-SpectreTable.tests.ps1 249ms (86ms|46ms)

Completing a single stage process ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100%


job 1 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100%
job 2 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100%

[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\progress\Invoke-SpectreCommandWithProgress.tests.ps1 2.3s (2.14s|41ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\progress\Invoke-SpectreCommandWithStatus.tests.ps1 1.25s (1.09s|41ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\prompts\Read-SpectreConfirm.tests.ps1 284ms (116ms|40ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\prompts\Read-SpectreMultiSelection.tests.ps1 286ms (112ms|48ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\prompts\Read-SpectreMultiSelectionGrouped.tests.ps1 364ms (204ms|42ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\prompts\Read-SpectrePause.tests.ps1 265ms (106ms|44ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\prompts\Read-SpectreSelection.tests.ps1 216ms (54ms|42ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\prompts\Read-SpectreText.tests.ps1 214ms (51ms|38ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\writing\Get-SpectreEscapedText.tests.ps1 171ms (8ms|42ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\writing\Write-SpectreCalender.tests.ps1 344ms (148ms|45ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\writing\Write-SpectreFigletText.tests.ps1 212ms (43ms|40ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\writing\Write-SpectreHost.tests.ps1 224ms (57ms|43ms)
[+] D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\writing\Write-SpectreRule.tests.ps1 175ms (18ms|40ms)
Tests completed in 10.83s
Tests Passed: 72, Failed: 0, Skipped: 0 NotRun: 0
Processing code coverage result.
Covered 43,21% / 75%. 1 560 analyzed Commands in 57 Files.

</details>

ShaunLawrie commented:
The publish won’t happen if the CI test fail, you could add a tag and
pester config to not run them in CI

On Fri, 19 Jan 2024 at 7:14 AM, trackd ***@***.***> wrote:

> @ShaunLawrie <https://github.com/ShaunLawrie> this is ready for
> prerelease.
>
> I ran through Start-SpectreDemo and the Pester tests, all looks good.
>
> CI tests failed but that's because they can only run 8bit so 24bit rgb is
> downgraded.
> Testlog Starting pwsh -noprofile and Running Pester tests...
>
> Starting discovery in 21 files.
> Discovery found 72 tests in 2.84s.
> Starting code coverage.
> Running tests.
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\formatting\ConvertTo-SpectreDecoration.tests.ps1
> 1.94s (1.33s|296ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\formatting\Format-SpectreBarChart.tests.ps1
> 653ms (443ms|86ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\formatting\Format-SpectreBreakdownChart.tests.ps1
> 369ms (167ms|63ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\formatting\Format-SpectreJson.tests.ps1
> 302ms (127ms|46ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\formatting\Format-SpectrePanel.tests.ps1
> 215ms (57ms|40ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\formatting\Format-SpectreTable.tests.ps1
> 504ms (307ms|69ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\formatting\Format-SpectreTree.tests.ps1
> 234ms (76ms|41ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\formatting\new_Format-SpectreTable.tests.ps1
> 249ms (86ms|46ms)
>
> Completing a single stage process ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
> 100%
>
> job 1 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100%
> job 2 ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100%
>
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\progress\Invoke-SpectreCommandWithProgress.tests.ps1
> 2.3s (2.14s|41ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\progress\Invoke-SpectreCommandWithStatus.tests.ps1
> 1.25s (1.09s|41ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\prompts\Read-SpectreConfirm.tests.ps1
> 284ms (116ms|40ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\prompts\Read-SpectreMultiSelection.tests.ps1
> 286ms (112ms|48ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\prompts\Read-SpectreMultiSelectionGrouped.tests.ps1
> 364ms (204ms|42ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\prompts\Read-SpectrePause.tests.ps1
> 265ms (106ms|44ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\prompts\Read-SpectreSelection.tests.ps1
> 216ms (54ms|42ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\prompts\Read-SpectreText.tests.ps1
> 214ms (51ms|38ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\writing\Get-SpectreEscapedText.tests.ps1
> 171ms (8ms|42ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\writing\Write-SpectreCalender.tests.ps1
> 344ms (148ms|45ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\writing\Write-SpectreFigletText.tests.ps1
> 212ms (43ms|40ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\writing\Write-SpectreHost.tests.ps1
> 224ms (57ms|43ms)
> [+]
> D:\Code\PwshSpectreConsole\PwshSpectreConsole.Tests\writing\Write-SpectreRule.tests.ps1
> 175ms (18ms|40ms)
> Tests completed in 10.83s
> Tests Passed: 72, Failed: 0, Skipped: 0 NotRun: 0
> Processing code coverage result.
> Covered 43,21% / 75%. 1 560 analyzed Commands in 57 Files.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ShaunLawrie/PwshSpectreConsole/pull/21#issuecomment-1899422192>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADEMYIRCW23TTO7RZQH2P4TYPG3F3AVCNFSM6AAAAABBKXRFVKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOJZGQZDEMJZGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

trackd commented:
> The publish won’t happen if the CI test fail, you could add a tag and pester config to not run them in CI

I've added a filter called 'ExcludeCI'to the 24bit tests and changed one of the tests a tiny bit.
need to look into why `LastWriteTime` gets truncated in CI..
I'm guessing its the `width = 80` so would need to use a different command with formatting and less columns in CI.

I'm not getting the conflict with ` PwshSpectreConsole.Tests/formatting/Format-SpectreTable.tests.ps1 ` in my fork so it doesnt look like i can resolve it?

trackd commented:
>The publish won’t happen if the CI test fail, you could add a tag and pester config to not run them in CI

I think maybe i was a bit unclear earlier.

All CI tests pass now

So it should be able to merge now.

(sorry if you knew and just havnt had time, its no rush from me) 

just thought that maybe my response might look weird on email as I had an expanding section.
